packages/client/src/client/Web/serviceWorkerRegistration.ts

```diff
@@ -53,51 +56,78 @@ export function register(config?: Config) {
         })
       } else {
         // Is not localhost. Just register service worker
-        registerValidSW(swUrl, config)
+        registerServiceWorker(swUrl, config)
       }
     })
   }
 }
 
-function registerValidSW(swUrl: string, config?: Config) {
+function addStateChangeHandler(
+  serviceWorkerInstance: ServiceWorker,
+  registration: ServiceWorkerRegistration,
+  registrationConfig: Config | undefined,
+) {
+  serviceWorkerInstance.onstatechange = () => {
+    console.debug(
+      "'installing' SW changed state to: ",
+      serviceWorkerInstance.state,
+    )
+    if (serviceWorkerInstance.state === "installed") {
+      if (navigator.serviceWorker.controller) {
+        // At this point, the updated precached content has been fetched,
+        // but the previous service worker will still serve the older
+        // content until all client tabs are closed.
+        console.debug(
+          "New content is available and will be used when all " +
+            "tabs for this page are closed. See https://cra.link/PWA.",
+        )
+        serviceWorkerInstance.postMessage({ type: "SKIP_WAITING" })
```

Comment:
just do the skip at the point of discovering we have an "installed" and waiting SW

Reply:
This might be the second time we are calling `SKIP_WAITING` , please see line 110 to 118 of this file, though I know this is redundant in terms of SW behavior if we call `SKIP_WAITING`.